soc/arm/renesas_rcar/common/pinctrl_rcar.h

```diff
@@ -0,0 +1,127 @@
+/** Alternate function for the pin is requested */
+#define RCAR_PIN_FLAGS_FUNC_SET BIT(3)
+
+#define RCAR_PIN_PULL_UP      (RCAR_PIN_FLAGS_PULL_SET | RCAR_PIN_FLAGS_PUEN | RCAR_PIN_FLAGS_PUD)
+#define RCAR_PIN_PULL_DOWN    (RCAR_PIN_FLAGS_PULL_SET | RCAR_PIN_FLAGS_PUEN)
+#define RCAR_PIN_PULL_DISABLE  RCAR_PIN_FLAGS_PULL_SET
+
+/** Type for R-Car pin. */
+typedef struct pinctrl_soc_pin {
+	uint16_t pin;
+	struct rcar_pin_func func;
+	uint8_t flags;
+	uint8_t drive_strength;
+} pinctrl_soc_pin_t;
+
+#define RCAR_IPSR(node_id) DT_PROP_BY_IDX(node_id, pin, 1)
+#define RCAR_HAS_IPSR(node_id) DT_PROP_HAS_IDX(node_id, pin, 1)
+
+/* Offsets are defined in dt-bindings pinctrl-rcar-common.h */
+#define RCAR_PIN_FUNC(node_id)			       \
+	{					       \
+		((RCAR_IPSR(node_id) >> 10U) & 0x1FU), \
+		((RCAR_IPSR(node_id) >> 4U) & 0x1FU),  \
+		((RCAR_IPSR(node_id) & 0xFU))	       \
+	}
+
+#define RCAR_PIN_FLAGS(node_id)						       \
+	DT_PROP(node_id, bias_pull_up)   * RCAR_PIN_PULL_UP |		       \
+	DT_PROP(node_id, bias_pull_down) * RCAR_PIN_PULL_DOWN |		       \
+	DT_PROP(node_id, bias_disable)   * RCAR_PIN_PULL_DISABLE |	       \
+	RCAR_HAS_IPSR(node_id) * RCAR_PIN_FLAGS_FUNC_SET
+
+#define RCAR_DT_PIN(node_id)						       \
+	{								       \
+		.pin = DT_PROP_BY_IDX(node_id, pin, 0),			       \
+		.func = COND_CODE_1(RCAR_HAS_IPSR(node_id),		       \
+				    (RCAR_PIN_FUNC(node_id)), (0)),	       \
+		.flags = RCAR_PIN_FLAGS(node_id),			       \
+		.drive_strength =					       \
+			COND_CODE_1(DT_NODE_HAS_PROP(node_id, drive_strength), \
+				    (DT_PROP(node_id, drive_strength)), (0)),  \
+	},
+
+/**
+ * @brief Utility macro to initialize each pin.
+ *
+ * @param node_id Node identifier.
+ * @param state_prop State property name.
+ * @param idx State property entry index.
+ */
+#define Z_PINCTRL_STATE_PIN_INIT(node_id, state_prop, idx) \
+	RCAR_DT_PIN(DT_PROP_BY_IDX(node_id, state_prop, idx))
+
+/**
+ * @brief Utility macro to initialize state pins contained in a given property.
+ *
+ * @param node_id Node identifier.
+ * @param prop Property name describing state pins.
+ */
+#define Z_PINCTRL_STATE_PINS_INIT(node_id, prop) \
+	{ DT_FOREACH_PROP_ELEM(node_id, prop, Z_PINCTRL_STATE_PIN_INIT) }
+
+struct pfc_drive_reg_field {
+	uint16_t pin;
+	uint8_t offset;
+	uint8_t size;
+};
+
+struct pfc_drive_reg {
+	uint32_t reg;
+	const struct pfc_drive_reg_field fields[8];
+};
+
+struct pfc_bias_reg {
+	uint32_t puen;		/** Pull-enable or pull-up control register */
+	uint32_t pud;		/** Pull-up/down or pull-down control register */
+	const uint16_t pins[32];
+};
+
+const struct pfc_bias_reg *pfc_rcar_get_bias_regs(void);
+const struct pfc_drive_reg *pfc_rcar_get_drive_regs(void);
+
+/**
+ * @brief set the register index for a given pin
+ *
+ * @param the pin
+ * @param pointer for the resulting register index
+ * @return 0 if the register index is found, negative
+ * errno otherwise.
+ */
+int pfc_rcar_get_reg_index(uint8_t pin, uint8_t *reg_index);
```

Comment:
this should not be part of pinctrl_soc, as it's something needed internally by drivers. please, create a new internal header so that drivers/pinctrl.h is not polluted with unnecessary stuff.

Reply:
Hi @gmarull ,
Thanks for your comment.
If my understanding is fine, you want me to move these common declaration to a new file here `zephyr/include/zephyr/drivers/pinctrl/pinctrl_rcar_common.h` ?

Reply:
Hi @gmarull, I'm currently reworking this PR to merge some changes with #60252 , can you confirm that I understood well your remark before I proceed ? Thanks

Reply:
What I mean is that this should only be exposed to code requiring it. An internal header is fine, but decoupled from pinctrl_soc.h

Reply:
Fixed it, please tel me if it's ok now !